### ISSUE TITLE
Rule proposal: `no-recursion`

### ISSUE THREAD

**Issue by fisker** [sindresorhus/eslint-plugin-unicorn#1078]
Disallow simply recursive function.

## Fail

```js
function foo(bar) {
  if(bar.baz) {
    return foo(bar.baz);
  }

  return bar;
}
```

## Pass

```js
function foo(bar) {
  while (bar.baz) {
    bar = bar.baz:
  }

  return bar;
}
```

```js
function foo(bar) {
  for (; bar.baz; bar = bar.baz) {}

  return bar;
}
```

```js
function foo(bar) {
  if(Array.isArray(bar)) {
    return bar.map(baz => foo(bar));
  }

  return bar;
}
```

I'm not sure how to define "simple".

**Comment by sindresorhus** [comment]
Can you include a reasoning in the proposal?

**Comment by fisker** [comment]
Recursive function may easily occur a "Maximum Callstack size exceeded" error, and it makes the error stack longer.

**Comment by sindresorhus** [comment]
I think they're also usually slightly slower as JS doesn't have [TCO](https://stackoverflow.com/a/37224563/64949).

**Comment by silverwind** [comment]
Recursion is necessary for some algorithms, one can not simply forbid it generally. I'd maybe call to rule `no-usless-recursion` but detecting which cases are useless will probably not be easy.

**Comment by sindresorhus** [comment]
I agree, it should be named `no-useless-recursion` and should only target some simple cases that can use loops instead.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by nawaz-pasha** [comment]
Is this picked up for development? Any update here?
